File: frontend-ci/src/services/notification.service.ts
/**
 * Service pour les notifications
 * Correspond à gestion/services/notification_service.py
 */

import apiClient from './api';
import { Notification, SaveNotificationDTO, UpdateNotificationDTO, CreateResponse } from '../types';
import { TypeNotification } from '../types/enums';

const ENDPOINTS = {
  BASE: '/notifications/',
  DETAIL: (id: number) => `/notifications/${id}/`,
  BY_TYPE: (type: TypeNotification) => `/notifications/?type=${type}`,
  ENVOYER_TOUS: (id: number) => `/notifications/${id}/envoyer-tous/`,
  ENVOYER_USER: (id: number, userId: number) => `/notifications/${id}/envoyer-user/${userId}/`,
};

export const notificationService = {
  /**
   * Récupérer toutes les notifications
   */
  async getAll(): Promise<Notification[]> {
    const response = await apiClient.get<Notification[]>(ENDPOINTS.BASE);
    return response.data;
  },

  /**
   * Récupérer une notification par son ID
   */
  async getById(id: number): Promise<Notification> {
    const response = await apiClient.get<Notification>(ENDPOINTS.DETAIL(id));
    return response.data;
  },

  /**
   * Récupérer les notifications par type
   */
  async getByType(type: TypeNotification): Promise<Notification[]> {
    const response = await apiClient.get<Notification[]>(ENDPOINTS.BY_TYPE(type));
    return response.data;
  },

  /**
   * Créer une nouvelle notification
   */
  async create(data: SaveNotificationDTO): Promise<CreateResponse> {
    const response = await apiClient.post<CreateResponse>(ENDPOINTS.BASE, data);
    return response.data;
  },

  /**
   * Mettre à jour une notification
   */
  async update(id: number, data: UpdateNotificationDTO): Promise<Notification> {
    const response = await apiClient.put<Notification>(ENDPOINTS.DETAIL(id), data);
    return response.data;
  },

  /**
   * Supprimer une notification
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(ENDPOINTS.DETAIL(id));
  },

  /**
   * Envoyer une notification à tous les utilisateurs
   */
  async envoyerATous(id: number): Promise<Notification> {
    const response = await apiClient.post<Notification>(ENDPOINTS.ENVOYER_TOUS(id));
    return response.data;
  },

  /**
   * Envoyer une notification à un utilisateur spécifique
   */
  async envoyerAUser(id: number, userId: number): Promise<Notification> {
    const response = await apiClient.post<Notification>(ENDPOINTS.ENVOYER_USER(id, userId));
    return response.data;
  },
};

export default notificationService;

File: backend-ci/gestion/services/notification_service.py
from django.db import transaction
from django.shortcuts import get_object_or_404
from gestion.models.notification import Notification
from gestion.models.notification_user import NotificationUser
from gestion.models.user import User


class NotificationService:

    @staticmethod
    @transaction.atomic
    def create(data):
        """Créer une nouvelle notification"""
        notification = Notification.objects.create(
            titre=data['titre'],
            message=data['message'],
            type=data.get('type', 'INFO')
        )
        return notification

    @staticmethod
    def get_all():
        """Récupérer toutes les notifications"""
        return Notification.objects.all().order_by('-date_envoi')

    @staticmethod
    def get_by_id(notification_id):
        """Récupérer une notification par son ID"""
        return get_object_or_404(Notification, id=notification_id)

    @staticmethod
    def get_by_type(type_notification):
        """Récupérer les notifications par type"""
        return Notification.objects.filter(type=type_notification).order_by('-date_envoi')

    @staticmethod
    def update(notification, data):
        """Mettre à jour une notification"""
        for key, value in data.items():
            setattr(notification, key, value)
        notification.save()
        return notification

    @staticmethod
    def delete(notification):
        """Supprimer une notification"""
        notification.delete()

    @staticmethod
    @transaction.atomic
    def envoyer_a_tous(notification):
        """Envoyer une notification à tous les utilisateurs"""
        users = User.objects.filter(actif=True)
        notification_users = [
            NotificationUser(user=user, notification=notification)
            for user in users
        ]
        NotificationUser.objects.bulk_create(notification_users)
        notification.envoyee = True
        notification.save()
        return notification

    @staticmethod
    @transaction.atomic
    def envoyer_a_user(notification, user_id):
        """Envoyer une notification à un utilisateur spécifique"""
        user = get_object_or_404(User, id=user_id)
        NotificationUser.objects.create(user=user, notification=notification)
        return notification
